src/Http/Middleware/CspMiddleware.php

```diff
@@ -0,0 +1,87 @@
+<?php
+declare(strict_types=1);
+
+/**
+ * CakePHP(tm) : Rapid Development Framework (http://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @link          http://cakephp.org CakePHP(tm) Project
+ * @since         4.0.0
+ * @license       http://www.opensource.org/licenses/mit-license.php MIT License
+ */
+
+namespace Cake\Http\Middleware;
+
+use Cake\Core\Configure;
+use InvalidArgumentException;
+use ParagonIE\CSPBuilder\CSPBuilder;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+
+/**
+ * Content Security Policy Middleware
+ */
+class CspMiddleware implements MiddlewareInterface
+{
+    /**
+     * CSP Builder
+     *
+     * @var \ParagonIE\CSPBuilder\CSPBuilder $csp CSP Builder or config array
+     */
+    protected $csp;
+
+    /**
+     * Constructor
+     *
+     * @param \ParagonIE\CSPBuilder\CSPBuilder|array|null $csp CSP object or config array
+     * @throws \Exception
+     */
+    public function __construct($csp = null)
+    {
+        if ($csp === null) {
+            $cspConfig = (array)Configure::read('App.CSP');
```

Comment:
It's better to avoid directly reading from `Configure`. Users can just pass array to the constructor.

Reply:
I like the internal handling in case it wasnt provided. We should do this more often where input is optional, and keeps it more DRY.

Reply:
CSP would better be Csp as we dont do acronym uppercases in code and config.
But if we are dropping this config now, never mind.

Reply:
Passing the array to constructor is easy enough when adding the middleware to queue in your application class so no need to try reading from `Configure`. Things are less magical this way and improves testability of middleware as all data is explicitly injected.

Reply:
This way was copied from #11115 
But I can remove it of course